### ISSUE TITLE
Writing to local file system in distributed operation does not work properly

### ISSUE THREAD

**Issue by rmetzger** [stratosphere/stratosphere#286]
I have a task that runs on two cluster nodes.
The output is set to "file:///home/rmetzger/cmpDiscount/", the `FileDataSink` has a DOP of 64.

After the job has finished, on host1, there is a directory in `/home/rmetzger/` called `cmpDiscount/` with 32 files inside.
On host2 on the other hand, there is only one file in my home directory, called `cmpDiscount`. So only one of the 32 threads on host2 writes into the file system?

**Comment by StephanEwen** [comment]
Did the job produce any errors? Did the log mention failed operations?
Am 25.11.2013 11:21 schrieb "Robert Metzger" notifications@github.com:

**Comment by rmetzger** [comment]
With INFO logging level, there are no warnings, nothing fails.

**Comment by rmetzger** [comment]
Even though the file exists, I got this exception (I guess its the same bug)
 (Just for the one who is going to fix this)

```
21:05:43,844 ERROR eu.stratosphere.nephele.taskmanager.runtime.RuntimeTask       - java.lang.RuntimeException: Stream to output file could not be opened: Opening the file output stream failed: File file:/home/rmetzger/cmpAge does not exis
t.
        at eu.stratosphere.pact.generic.io.FileOutputFormat.open(FileOutputFormat.java:115)
        at eu.stratosphere.pact.common.io.RecordOutputFormat.open(RecordOutputFormat.java:167)
        at eu.stratosphere.pact.runtime.task.DataSinkTask.invoke(DataSinkTask.java:163)
        at eu.stratosphere.nephele.execution.RuntimeEnvironment.run(RuntimeEnvironment.java:344)
        at java.lang.Thread.run(Thread.java:679)
Caused by: java.io.IOException: Opening the file output stream failed: File file:/home/rmetzger/cmpAge does not exist.
        at eu.stratosphere.pact.generic.io.FileOutputFormat$OutputPathOpenThread.waitForCompletion(FileOutputFormat.java:202)
        at eu.stratosphere.pact.generic.io.FileOutputFormat.open(FileOutputFormat.java:112)
        ... 4 more
Caused by: java.io.FileNotFoundException: File file:/home/rmetzger/cmpAge does not exist.
        at eu.stratosphere.nephele.fs.file.LocalFileSystem.getFileStatus(LocalFileSystem.java:106)
        at eu.stratosphere.pact.generic.io.FileOutputFormat$OutputPathOpenThread.run(FileOutputFormat.java:162)
```

**Comment by StephanEwen** [comment]
Does the file (or directory) really not exist, or is the message wrong?

**Comment by rmetzger** [comment]
There is a file with exactly that name but since the dop on that node is higher than 1, there should be a directory.
So the exception message is wrong.
But I assume that the system creates the directory if it does not exist. (it does on the host where I start the job)

**Comment by StephanEwen** [comment]
Ah, there is your solution: The Directory is created by the master when the
outputs are prepared. They are not created by each workers.

We should change that ;-)

**Comment by StephanEwen** [comment]
Preparing the output happens during the JobGraph preparation and checking
phase

**Comment by fhueske** [comment]
I started with this one.

**Comment by fhueske** [comment]
Fixed, waiting for #257 to be merged.

**Comment by fhueske** [comment]
Fixed in #421.
